src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.CheckOrReachability.cs

```diff
@@ -0,0 +1,1084 @@
+                    candidateTypes.Free();
+                }
+                else
+                {
+                    narrowedType = resultRight.NarrowedType;
+                }
+
+                return narrowedType;
+            }
+
+            TypeSymbol? LeastSpecificType(ArrayBuilder<TypeSymbol> candidates)
+            {
+                CompoundUseSiteInfo<AssemblySymbol> useSiteInfo = CompoundUseSiteInfo<AssemblySymbol>.Discarded;
+                return Binder.LeastSpecificType(_conversions, candidates, ref useSiteInfo);
+            }
+
+            public override BoundNode? VisitNegatedPattern(BoundNegatedPattern node)
+            {
+                var savedNegated = _negated;
+                _negated = !_negated;
+                var result = this.Visit(node.Negated);
+                _negated = savedNegated;
+                return result;
+            }
+
+            public BoundPattern NegateIfNeeded(BoundPattern node)
+            {
+                if (!_negated)
+                {
+                    return node;
+                }
+
+                if (node is BoundNegatedPattern { Negated: var negated })
+                {
+                    return negated;
+                }
+
+                var result = new BoundNegatedPattern(node.Syntax, node, node.InputType, narrowedType: node.InputType);
+                if (node.WasCompilerGenerated)
+                {
+                    result.MakeCompilerGenerated();
+                }
+
+                return result;
+            }
+
+            public override BoundNode? VisitTypePattern(BoundTypePattern node)
+            {
+                return NegateIfNeeded(WithInputTypeCheckIfNeeded(node));
+            }
+
+            public override BoundNode? VisitConstantPattern(BoundConstantPattern node)
+            {
+                return NegateIfNeeded(WithInputTypeCheckIfNeeded(node));
+            }
+
+            public override BoundNode? VisitDiscardPattern(BoundDiscardPattern node)
+            {
+                return NegateIfNeeded(WithInputTypeCheckIfNeeded(node));
+            }
+
+            public override BoundNode? VisitRelationalPattern(BoundRelationalPattern node)
+            {
+                return NegateIfNeeded(WithInputTypeCheckIfNeeded(node));
+            }
+
+            private BoundPattern WithInputTypeCheckIfNeeded(BoundPattern pattern)
+            {
+                return WithInputTypeCheckIfNeeded(pattern, _inputType);
+            }
+
+            private static BoundPattern WithInputTypeCheckIfNeeded(BoundPattern pattern, TypeSymbol inputType)
+            {
+                // Produce `PatternInputType and pattern` given a new input type
+
+                if (pattern.InputType.Equals(inputType, TypeCompareKind.AllIgnoreOptions))
+                {
+                    return pattern;
+                }
+
+                BoundPattern typePattern = new BoundTypePattern(pattern.Syntax,
```

Comment:
>BoundTypePattern [](http://example.com/codeflow?start=47&length=16)

Assuming, a type pattern implies a `not null` test, wouldn't it be wrong to combine it with a `is null` pattern? And any other pattern that might succeed for a null value?